cli/install_ee/crawler.go

```diff
@@ -0,0 +1,220 @@
+				!isWrongOsLink && !isSHA256Link && !isSameLink {
+				downloadUrls = append(downloadUrls, hrefValue)
+			}
+		}
+	}
+}
+
+// collectBundleReferences crawls all urls from the passed bundle source, collects references
+// containing sdk bundles for the host system and returns them in a slice of strings.
+func collectBundleReferences(searchCtx SearchEECtx, baseUrl string,
+	credentials userCredentials) ([]string, error) {
+	bundleRefsQueue := &stringQueue{mtx: &sync.Mutex{}, buf: make([]string, 0)}
+	prefixQueue := &stringQueue{mtx: &sync.Mutex{}, buf: make([]string, 0)}
+
+	osType, err := util.GetOs()
+	if err != nil {
+		return nil, err
+	}
+	if osType == util.OsMacos {
+		prefixQueue.InsertBatch([]string{eeReleasePrefix, eeDebugPrefix, eeDevPrefix,
+			eeMacosOldPrefix})
+	} else {
+		prefixQueue.Insert(eeCommonPrefix)
+	}
+
+	eeUrl, err := url.Parse(baseUrl)
+	if err != nil {
+		return nil, err
+	}
+
+	vProcCount := runtime.GOMAXPROCS(0) * 5
+
+	// errChan is a channel for receiving errors from crawling goroutines.
+	errChan := make(chan error, vProcCount)
+
+	// freeAtomicSl is a slice of atomic values for controlling the state of workers.
+	freeAtomicSl := make([]atomic.Bool, vProcCount)
+	for i := range freeAtomicSl {
+		freeAtomicSl[i] = atomic.NewBool(true)
+	}
+
+	// prefixChanSl is a slice of channels with relative urls that is used by
+	// workers goroutines for receiving a new url for crawling.
+	prefixChanSl := make([]chan string, vProcCount)
+	for i := range prefixChanSl {
+		prefixChanSl[i] = make(chan string, 1)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	for i := 0; i < vProcCount; i++ {
+		go crawlWorker(ctx, searchCtx, prefixChanSl[i], prefixQueue, bundleRefsQueue,
+			eeUrl.String(), errChan, freeAtomicSl[i], credentials)
+	}
+
+	// This loop is needed for managing workers. Any worker puts found
+	// references to the prefix queue and waits for a new prefix to be
+	// received from the prefix channel. This loop implements the logic
+	// of moving prefixes from the queue to the channels of free workers.
+	// If all workers have done their work and there is no more prefixes
+	// in the queue, the loop will stop all the goroutines by canceling
+	// the context.
+	for {
+		select {
+		case err := <-errChan:
+			cancel()
+			return nil, err
+		default:
+		}
+		countFree := 0
+		for i := range freeAtomicSl {
+			if freeAtomicSl[i].IsSet() {
+				prefix, err := prefixQueue.Pop()
+				if err != nil {
+					countFree++
+					if countFree == vProcCount {
+						cancel()
+						return bundleRefsQueue.GetAll()
+					}
+				} else {
+					freeAtomicSl[i].Unset()
+					prefixChanSl[i] <- prefix
+				}
+			}
+		}
+	}
+}
+
+// crawlWorker is a worker function that gets a new relative url from the prefix channel,
+// scans it and fills a queue with found prefixes and a queue with found bundle references.
+func crawlWorker(ctx context.Context, searchCtx SearchEECtx, prefixChan chan string,
```

Comment:
Do we really need worker "threads"? Can we just invoke goroutine for each found page or package? So we don't need to use queue, atomics and other sync stuff. If we need to limit simultaneous goroutines count, buferred channel can be used like this:
```go
limitChan <- struct{}{}
go doJobWIthUri()
<-limitChan
```

Additionally. I think it is useful to have crawler api in common place, so it can be used later for parallel parsing of the pages. Looks like, now it is tightly bound with EE. For example, crawler should not know if it is a search or some other processing.